capactio/capact
# Validate parent nodes in manifests

## Description

Changes proposed in this pull request:

- Validate parent nodes in manifests
   Check if item under `requires` section is a concrete Type. If yes, it needs to be attached to the parent node. For example:
    ```yaml
    requires:
      cap.core.type.platform:
       oneOf:
         - name: cap.type.platform.cloud-foundry # this MUST be attached to `cap.core.type.platform`
           revision: 0.1.0
    ```
- Validate that **additionalRefs** always point to abstract node (parent node) not to a concrete Type.
- In `fsvalidator_test.go` I added only simple cases. Corner cases were extracted to `json_remote_implementation_test.go`.
- Replace “hard-coded” `ListTypeRefRevisionsJSONSchemas` with generic `ListTypes` function in Hub client
- Introduce a single FakeHubCli with related "fixtures" and remove all copies inside our code
- Detects situation where the new revision is not attached to previous parent node. Introduced here: https://github.com/capactio/capact/pull/621/commits/8b9ff424e5738748a52ce0ccf80f03e6c4a0e03f
  - it complicates our logic but give us more robustness
  - I didn't have more time to try to make it more readable but current state IMO is enough


## Testing


1. Build CLI: `make build-tool-cli`
1. Pick some Implementation manifest and add change `requires` property:
    ```yaml
        cap.core.type.platform:
          oneOf:
            - name: kubernetes
              revision: 0.1.0
            - name: cap.type.platform.cloud-foundry
              revision: 0.1.0
            - name: cap.type.aws.auth.credentials
              revision: 0.1.0
    ```

1. Execute validation against local manifests:
    NOTE: use the CLI from this PR
    ```bash
    capact manifest validate --server-side ./manifests/ --recursive
    ```

    you should see:
    ```bash
    Validating files in 5 concurrent jobs...
    - ✗ "manifests/implementation/mattermost/mattermost-team-edition/install.yaml":
        * RemoteImplementationValidator: Type cap.type.aws.auth.credentials is not attached to cap.core.type.platform abstract node
    ```

1. Change **additionalRefs** in  the `cap.type.platform.cloud-foundry` manifests from `cap.core.type.platform` to `cap.core.type.platform.kubernetes`

1. Execute validation one more time to see error with wrong parent node:
    ```bash
    - ✗ "manifests/type/platform/cloud-foundry.yaml":
        * RemoteTypeValidator: cap.core.type.platform.kubernetes cannot be used as parent node as it resolves to concrete Type
    ```

## Related issue(s)

- https://github.com/capactio/capact/issues/601
